### ISSUE TITLE
Tighter bounds on split sizes for multi-worker use cases

### ISSUE THREAD

**Issue by rwightman** [tensorflow/datasets#3119]
**Is your feature request related to a problem? Please describe.**
I'm having issues performing distributed validation in the case where I have multiple workers with TFDS.

Neither the tf.distribute.InputContext(num_input_pipelines=) or subsplit approach gives me what I'm hoping to achieve.

For a concrete scenario, I'm trying to perform validation across 8 worker nodes (separate processes) w/ TFDS on ImageNet. For this to work well, without extra code, masking, batch size counts, etc I need very even splits across each node. I +/- 1 is okay, I can repeat a sample if needed and the impact is minor but I see much greater than that.

For InputContext(num_input_pipelines=8) I get this # samples per worker:
6248
6248
6248
6256
6248
6256
6248
6248

For tfds.even_splits(split, 8) I get:
6000
6000
6000
6000
6500
6500
6500
6500 

Is this a bug? I was actually expecting subsplits/even splits to be better, because I assumed it worked at a finer granularity than InputContext. For InputContext it matches my expectations and appears to be splitting based on the sharded files (781 or 782 images per shard).

I'm ideally looking for 6250 per worker since 50,000K for ImageNet is evenly divisible. If it's not evenly divisible I'd expect __no more__ than `ceil(num_examples / num_workers)` per worker and I can pad the workers where it's short.

In my current use case I really only care about this for validation with shuffling disabled. For training, I'm already tossing partial last batches, it's shuffled, has minimal impact. 

**Describe the solution you'd like**

The ability to specify splits with tighter bounds such that no worker has more than ceil(num_examples / num_workers) examples. It is okay if this is limited to the case where shuffled_files=False.

**Describe alternatives you've considered**
Tracking # samples per batch / num batches per process and writing some complicated masking/distributed reduction logic for metric calculation. Or take the speed hit and always validate on rank==0 :(

**Comment by rwightman** [comment]
Silly me, should have poked around a bit longer. Looks like I can just split on the record indices themselves and ignore the whole % / even_splits API. Any reason there isn't an even_split equivalent that is based on indices?

One more question related to this if anyone happens to read before I close. If I want to manually partion a subsplit (ie I want to divide `train[90%:]` into 8 subsets with indices specified), I clearly can't layer subsplits, as in do train[90%:][start:end] for my manual sharding.

So, I thought of doing something like below, go back to absolute indices in the 'train' split. Is there any way to calculate the range of indices in the original split for a subsplit specified by % so that I end up with the exact same samples?
```
split = 'train[90%:]'
num_samples = builder.info.splits[split].num_examples
train_offset = ??? # <--- need offset into 'train' for train[90%:]
manual_subsplit = even_split_indices('train', num_splits, num_samples, train_offset)  
ds = builder.as_dataset(split=manual_subsplit...)
```

**Comment by rwightman** [comment]
This additions in #3431 look like a complete re-write of even_splits and fixed this problem in exactly the way it needed to be done despite being completed unrelated to the commit message :)
